main.go

```diff
@@ -368,6 +371,64 @@ func makeHorizontalPodAutoscalerChanges(kubeClient *k8s.Client, hpa *autoscaling
 	return status, nil
 }
 
+// Returns what the minimum pod count should be based on the Prometheus query specified
+// If the Prometheus query is not specified, it returns 0
+func getMinPodCountBasedOnPrometheusQuery(kubeClient *k8s.Client, hpa *autoscalingv1.HorizontalPodAutoscaler, initiator string, desiredState HPAScalerState) (minPodCount int32, requestRate float64, err error) {
+	minPodCount = 0
+	requestRate = 0
+
+	if len(desiredState.PrometheusQuery) > 0 && desiredState.RequestsPerReplica > 0 {
+		// get request rate with prometheus query
+		// http://prometheus.production.svc/api/v1/query?query=sum%28rate%28nginx_http_requests_total%7Bhost%21~%22%5E%28%3F%3A%5B0-9.%5D%2B%29%24%22%2Clocation%3D%22%40searchfareapi_gcloud%22%7D%5B10m%5D%29%29%20by%20%28location%29
+		prometheusQueryURL := fmt.Sprintf("%v/api/v1/query?query=%v", desiredState.PrometheusServerURL, url.QueryEscape(desiredState.PrometheusQuery))
+		resp, err := pester.Get(prometheusQueryURL)
+		if err != nil {
+			log.Error().Err(err).Msgf("Executing prometheus query for hpa %v in namespace %v failed", *hpa.Metadata.Name, *hpa.Metadata.Namespace)
+			return 0, 0, err
+		}
+
+		defer resp.Body.Close()
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Error().Err(err).Msgf("Reading prometheus query response body for hpa %v in namespace %v failed", *hpa.Metadata.Name, *hpa.Metadata.Namespace)
+			return 0, 0, err
+		}
+
+		queryResponse, err := UnmarshalPrometheusQueryResponse(body)
+		if err != nil {
+			log.Error().Err(err).Msgf("Unmarshalling prometheus query response body for hpa %v in namespace %v failed", *hpa.Metadata.Name, *hpa.Metadata.Namespace)
+			return 0, 0, err
+		}
+
+		requestRate, err = queryResponse.GetRequestRate()
+		if err != nil {
+			log.Error().Err(err).Msgf("Retrieving request rate from query response body for hpa %v in namespace %v failed", *hpa.Metadata.Name, *hpa.Metadata.Namespace)
+			return 0, 0, err
+		}
+
+		// calculate target # of replicas
+		minPodCount = int32(math.Ceil(desiredState.Delta + requestRate/desiredState.RequestsPerReplica))
+	}
+
+	return minPodCount, requestRate, nil
+}
+
+// Returns what the minimum pod count should be based on the current pod count and the maximum scale down ratio
+func getMinPodCountBasedOnCurrentPodCount(kubeClient *k8s.Client, hpa *autoscalingv1.HorizontalPodAutoscaler, initiator string, desiredState HPAScalerState) (podCount int32) {
```

Comment:
If you use this without a prometheus query it just sets a percentage of the current number of replicas, right? Let's say the ScaleDownMaxRatio is 0.1 then it will result in the minimum being set to 0.9 times the actualNumberOfReplicas.

Reply:
Yep, exactly.